pkg/workload/kv: refactor kv Zipf RNG cycle and fix various bugs and races

sean- opened the pull request:
* Update `workload kv --zipfian` to add `--zipfian-v` and `--zipfian-s` CLI flags to change rand.Zipf's data distribution

* Replace custom zipfian implementation with Go's built-in rand.Zipf.  Initialize the Zipf RNG once, but implemented a custom rand.Source that iterates through a sequence of `--cycle-length`

* Added `--zipfian-mode={static,sliding}` to change `--zipfian`'s behavior.  The default `static` mode maintains the current behavior, where the key sequence wraps after a fixed `--cycle-length`, creating persistent hotspots (useful for testing range split operations). The new `sliding` mode implements a sliding window mechanism that rotates the key range over time, simulating evolving customer workloads more realistically.

Release note: Update `workload kv --zipfian` to add `--zipfian-v` and --zipfian-s` CLI flags to change the data distribution
Epic: none

sean- committed:
pkg/workload/kv: refactor kv Zipf RNG cycle

* Replace custom zipfian implementation with Go's built-in rand.Zipf

* Refactored sources don't immediately request a recently written value

Epic: none

Release note: None

sean- committed:
workload: add zipfian S and V parameters to workload kv --zipfian

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142733)
<!-- Reviewable:end -->

sean- committed:
pkg/workload: Add --zipfian-mode={static,sliding} to modify --zipfian behavior

This commit introduces a new command-line flag, --zipfian-mode, to allow users
to choose between two modes for Zipfian key allocation. The default "static"
mode maintains the current behavior, where the key sequence wraps after a fixed
cycle-length, creating persistent hotspots (useful for testing range split
operations). The new "sliding" mode implements a sliding window mechanism that
rotates the key range over time, simulating evolving customer workloads more
realistically.

Epic: none
Release note (cli change): Added --zipfian-mode flag to support both static and
sliding Zipfian key generation, which more closely mirrors customer workloads.

DarrylWong reviewed (commented):
It would be nice to clean up the commits here, would make it nicer on the reviewers. More concretely:

1. Can you merge and squash relevant commits? e.g. it's harder to review a commit that introduces a bug that is fixed 5 commits later or try and understand a function that isn't documented until a future commit.
2. This PR is really two unrelated changes, zipf refactoring and the race condition fix. They are both non trivial enough that I'd prefer them to be separate PRs. It will be easier to review the changes individually and will prevent us having to back out all the changes if theres a bug.

sean- commented:
> Can you merge and squash relevant commits? e.g. it's harder to review a commit that introduces a bug that is fixed 5 commits later or try and understand a function that isn't documented until a future commit.
Done.  The second commit in the series contains three distinct things: various `--ramp` fixes, the switch to `math/rand/v2`, and the the `--zipfian-s` and `--zipfian-v` flags.  Given how that all came to be, I didn't tease the individual patches apart when rebasing this branch and just left it alone given the ROI on splitting things apart vs creating individual PRs.

sean- commented:
This is a resubmission of #142258.

Commit:
Merge branch 'master' into pkg-workload-omnibus-fixes
